### PR TITLE
Notification for pending release

### LINKED ISSUE #14: Improvements
- create and test npm package
- create workflow template
- check and test error handling
- improve readme
- create npm pckg & test
- extend by having an exclude for some file/dir changes

### PR DESCRIPTION
**Pull request by sameer-coder**
This github action is fully functional now and creates an issue with a list of pending commits that are older than X days.

Closes #14

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,26 +5,30 @@
...
-    "lint": "eslint ."
+    "test": "jest --coverage"
   },
   "repository": {
     "type": "git",
-    "url": "git+https://github.com/estherixz/github-action-notify-release.git"
+    "url": "git+https://github.com/nearform/github-action-notify-release.git"
   },
-  "author": "Esther Saimpou",
+  "author": "Sameer Srivastava",
```
Comment (simoneb) [inline package.json:14]: don't change the author, add yourself as a contributor instead :)

File: package.json [hunk 2]
```diff
@@ -5,26 +5,30 @@
...
   },
   "devDependencies": {
     "@vercel/ncc": "^0.28.3",
-    "eslint": "^7.24.0"
+    "eslint": "^7.25.0",
+    "eslint-config-airbnb": "^18.2.1",
+    "eslint-config-airbnb-base": "^14.2.1",
+    "eslint-plugin-import": "^2.22.1",
+    "jest": "^26.6.3",
+    "tap": "^15.0.5"
```
Comment (simoneb) [inline package.json:32]: remove the one you're not using

File: dist/index.js [hunk 3]
```diff
@@ -1,7 +1,7 @@
 /******/ (() => { // webpackBootstrap
 /******/ 	var __webpack_modules__ = ({
 
-/***/ 7351:
```
Comment (simoneb) [inline dist/index.js:4]: why did you remove this file? is the dist folder not necessary?

File: package.json [hunk 4]
```diff
@@ -5,26 +5,30 @@
   "main": "dist/index.js",
```
Comment (simoneb) [inline package.json:5]: to the previous point about the dist folder. if we bundle the action we do need the dist folder to be in the repo

File: src/utils.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+const core = require('@actions/core');
+const github = require('@actions/github');
+const { logInfo } = require('./log');
+
+async function createIssue(token, issueTitle, issueBody) {
+  try {
+    const octokit = github.getOctokit(token);
+
+    logInfo(issueBody);
```
Comment (simoneb) [inline src/utils.js:9]: this doesn't look like something to be logged with level info. do we need to log this in the first place?

File: src/utils.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+    return await octokit.issues.create({
+      ...github.context.repo,
+      title: issueTitle,
+      body: issueBody,
+    });
+  } catch (error) {
+    core.setFailed(error.message);
```
Comment (simoneb) [inline src/utils.js:17]: do we need this? I mean if an exception is thrown won't the action fail without us having to handle errors?

File: test/release.spec.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+const token = 'dummytoken'
+
+jest.mock('@actions/github', () => ({
+  getOctokit: jest.fn(),
+  context: { repo: { owner: "sameer", repo: "testrepo" } }
+}))
+
+test("Gets the latest release of the repository", async () => {
+  getOctokit.mockImplementation(() => {
+    return { request: async () => mock_allReleases }
```
Comment (simoneb) [inline test/release.spec.js:22]: if you're not declaring this behavior during mock setup, you don't need the mock prefix

File: test/release.spec.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+  });
+  const response = await getLatestRelease(token);
```
Comment (simoneb) [inline test/release.spec.js:24]: naming: response is misleading. this function returns the latest release, so I'd call it latestRelease

File: src/release.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  const { owner, repo } = github.context.repo;
+
+  // TODO: extract this
+  const allCommitsResp = await octokit.request('GET /repos/{owner}/{repo}/commits', {
+    owner,
+    repo,
+  });
+
+  if (!allCommitsResp || !allCommitsResp.data || !allCommitsResp.data.length) throw new Error('Error fetching commits');
+  if (!latestRelease || !latestRelease.created_at) throw new Error('Latest release doesnt have a created_at date');
```
Comment (simoneb) [inline src/release.js:30]: formatting: use curly braces and go to new line please

File: src/release.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+  // eslint-disable-next-line no-param-reassign
+  if (!daysToIgnore) daysToIgnore = 0;
```
Comment (simoneb) [inline src/release.js:32]: avoid ignoring eslint rules. this function should get a valid value of daysToIgnore, it should not do this logic. This logic should live where the parameter is read in the first place, meaning the caller of this function

File: src/release.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+  const unreleasedCommits = [];
+  const lastReleaseDate = new Date(latestRelease.created_at).getTime();
+  let staleDate = new Date().getTime();
+  if (daysToIgnore > 0) {
+    staleDate = new Date().getTime() - (daysToIgnore * 24 * 60 * 60 * 1000);
+  }
```
Comment (simoneb) [inline src/release.js:39]: you don't need special handling for > 0. if it's 0, this formula will still apply and will return the current time

File: src/release.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    const commitDate = new Date(commit.commit.author.date).getTime();
+    if (lastReleaseDate < commitDate && commitDate < staleDate) {
+      unreleasedCommits.push({
+        message: commit.commit.message,
+        author: commit.commit.author.name,
+        date: commitDate,
+        url: commit.url,
+      });
+    }
+  }
```
Comment (simoneb) [inline src/release.js:51]: this is a big no: we can't possibly iterate all commits in a repo. we need to filter them when we request them from the API if we want to include them in the issue we create. if we can't filter them at the GH API level we should not include them in the first place. filtering them means asking for the commits after the last release date to the API, instead of doing the filter in code.

File: src/index.js [hunk 13]
```diff
@@ -1,36 +1,46 @@
...
 
-    const daysSinceRelease = differenceInDays(lastRepoUpdate, lastReleaseDate)
+    logInfo(`Latest release - name:${latestRelease.name}, created:${latestRelease.created_at},
+Tag:${latestRelease.tag_name}, author:${latestRelease.author.login}`);
 
-    logInfo(`${daysSinceRelease} Days since last release`)
+    const unreleasedCommits = await getUnreleasedCommits(token, latestRelease, daysToIgnore);
 
-    const shouldCreateIssue = daysToStaleRelease < daysSinceRelease
+    logInfo(JSON.stringify(unreleasedCommits));
```
Comment (simoneb) [inline src/index.js:24]: let's not log all this stuff, no need to clutter the output of the action

File: src/index.js [hunk 14]
```diff
@@ -1,36 +1,46 @@
...
-const { createIssue } = require('./createIssue')
+const core = require('@actions/core');
+const { logInfo } = require('./log');
+const { getLatestRelease, getUnreleasedCommits } = require('./release');
+const { createIssue } = require('./utils');
 
 async function run() {
   try {
-    const daysToStaleRelease = core.getInput('days-to-stale-release')
+    logInfo('========Starting to run the stale release github action ============');
```
Comment (simoneb) [inline src/index.js:8]: this is too much logging that is not needed, see other places as well

File: src/index.js [hunk 15]
```diff
@@ -1,36 +1,46 @@
...
 
-    const { lastReleaseDate, npmError, npmPckNotFound } = await getLastReleaseDate()
+    const token = core.getInput('github-token', { required: true });
 
-    if (npmPckNotFound) return logError('npm package not found')
+    const daysToIgnore = core.getInput('days-to-ignore');
```
Comment (simoneb) [inline src/index.js:12]: is this required? if not, it should have a default value. see other comment about this too.

File: src/index.js [hunk 16]
```diff
@@ -1,36 +1,46 @@
...
 
-    if (npmError) return logError(npmError)
+    logInfo(`Days since last release: ${daysToIgnore}`);
+    logInfo('Fetching latest release......');
```
Comment (simoneb) [inline src/index.js:15]: too much logging. this is debugging information

File: src/index.js [hunk 17]
```diff
@@ -1,36 +1,46 @@
...
-    const shouldCreateIssue = daysToStaleRelease < daysSinceRelease
+    logInfo(JSON.stringify(unreleasedCommits));
 
-    if (shouldCreateIssue) {
-      createIssue(daysSinceRelease)
+    if (unreleasedCommits.length) {
+      let commitStr = '';
+      for (const commit of unreleasedCommits) {
+        commitStr += `Issue: ${commit.message}<br>Author: ${commit.author}<br><br>`;
+      }
```
Comment (simoneb) [inline src/index.js:30]: use markdown, not html (avoid <br> please). also, just map the commits to an array and join them instead of concatenating string would be more elegant in my opinion.

File: src/release.js [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+  const { owner, repo } = github.context.repo;
+
+  const allReleasesResp = await octokit.request('GET /repos/{owner}/{repo}/releases', {
+    owner,
+    repo,
+  });
+
+  const latestRelease = (allReleasesResp && allReleasesResp.data && allReleasesResp.data.length)
+    ? allReleasesResp.data[0] : null;
+  if (!latestRelease) throw new Error('Cannot find the latest release');
```
Comment (simoneb) [inline src/release.js:14]: use curly braces and new line

File: src/release.js [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+async function getLatestRelease(token) {
+  const octokit = github.getOctokit(token);
+  const { owner, repo } = github.context.repo;
+
+  const allReleasesResp = await octokit.request('GET /repos/{owner}/{repo}/releases', {
+    owner,
+    repo,
+  });
+
+  const latestRelease = (allReleasesResp && allReleasesResp.data && allReleasesResp.data.length)
```
Comment (simoneb) [inline src/release.js:12]: please check the docs about what you should expect here. this condition may be simplified as you can probably assume that you get something from the octokit request

File: src/release.js (hunk 9 above)
Comment (simoneb) [inline src/release.js:30]: same thing as above, these conditions can probably be simplified if you can make assumptions about the return values of the octokit apis

**Review by simoneb — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 15 above)
Comment (sameer-coder) [reply src/index.js:12]: It takes the input from the user. Also, I have now included a default value in parameters in case the user doesn't provide one

Comment (simoneb) [reply src/index.js:12]: why don't we simply make it a required input? or non-required with a default is also fine. in the second case, let's document what is the default, of course

Comment (sameer-coder) [reply src/index.js:12]: I have made it optional with default being 0 (documented in the readme)

Comment (simoneb) [reply src/index.js:10]: do you think 0 is a sensible default?

File: .eslintrc.js [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+    'airbnb-base',
+  ],
+  parserOptions: {
+    ecmaVersion: 12,
+  },
+  rules: {
+    'no-restricted-syntax': 'off',
+    'prefer-template': 'off',
+    'eslint-disable strict': 'off',
+    'eslint-env': 'jest'
```
Comment (simoneb) [inline .eslintrc.js:16]: are you sure this rule is right? I can't find any reference to it. I always did it as described here https://stackoverflow.com/a/57328996/32093

File: .github/workflows/notify-release.yml [hunk 21]
```diff
@@ -9,12 +9,9 @@ jobs:
...
-      - uses: actions/setup-node@v2
-        with:
-          node-version: ${{ matrix.node-version }}
       - name: Check if there's a release pending
         id: notify
-        uses: estherixz/github-action-notify-release@main
+        uses: nearform/github-action-notify-release@main
         with:
           github-token: ${{ secrets.GITHUB_TOKEN }}
+          days-to-ignore: 0
```
Comment (simoneb) [inline .github/workflows/notify-release.yml:17]: let's rename this to something more telling. `stale-days` sounds more meaningful. also reflect in docs and elsewhere

File: .github/workflows/notify-release.yml [hunk 22]
```diff
@@ -9,12 +9,9 @@ jobs:
   setup:
```
Comment (simoneb) [inline .github/workflows/notify-release.yml:9]: please name this workflow after the file name. either notify-pending-release or notify-release, but avoid inconsistent naming.

in addition to this workflow, we need a workflow which runs eslint and tests automatically. inspire from how we do the same in other workflows

File: action.yml [hunk 23]
```diff
@@ -1,12 +1,12 @@
...
 description: 'Create an issue if there is a release pending'
 inputs:
   github-token:
     description: 'GitHub token'
     required: true
-  days-to-stale-release:
-    description: 'The number of days after which notification should trigger after last unreleased commit'
-    required: true
+  days-to-ignore:
+    description: 'Unreleased commits older than the no. of days mentioned in this setting shall be reported'
```
Comment (simoneb) [inline action.yml:8]: this description is not accurate because, as we discussed, because we're opening an issue if there are commits stale-days after the last release and we include all of them in the issue. the previous one was slightly better but still not completely accurate either

File: package.json [hunk 24]
```diff
@@ -5,26 +5,29 @@
...
   "scripts": {
     "build": "ncc build src --license licenses.txt",
-    "lint": "eslint ."
+    "test": "jest --coverage"
```
[the remark this answers is not in the document]
Comment (simoneb) [reply package.json:8]: the way you added husky is not correct. please check out the docs. also, for jest please use a configuration file instead of providing cli flags and configure so that the required coverage is 100%.

File: package.json [hunk 25]
```diff
@@ -5,26 +5,35 @@
   "main": "dist/index.js",
   "scripts": {
     "build": "ncc build src --license licenses.txt",
-    "lint": "eslint ."
```
Comment (simoneb) [inline package.json:8]: please restore the lint script, it should be ran as part of the new ci.yml workflow

File: package.json [hunk 26]
```diff
@@ -5,26 +5,35 @@
...
     "@actions/core": "^1.2.7",
-    "@actions/exec": "^1.0.4",
     "@actions/github": "^4.0.0",
     "date-fns": "^2.21.1"
   },
   "devDependencies": {
     "@vercel/ncc": "^0.28.3",
-    "eslint": "^7.24.0"
+    "eslint": "^7.25.0",
+    "eslint-config-airbnb": "^18.2.1",
```
Comment (simoneb) [inline package.json:28]: if we use the base config, is this one required?

File: src/index.js [hunk 27]
```diff
@@ -1,36 +1,57 @@
...
-    const lastRepoUpdate = await getLastRepoUpdate()
-
-    const daysSinceRelease = differenceInDays(lastRepoUpdate, lastReleaseDate)
-
-    logInfo(`${daysSinceRelease} Days since last release`)
-
-    const shouldCreateIssue = daysToStaleRelease < daysSinceRelease
+    const token = core.getInput('github-token', { required: true });
+    const daysInput = core.getInput('days-to-ignore');
+    const daysToIgnore = Number(daysInput);
```
Comment (simoneb) [inline src/index.js:10]: avoid declaring the daysInput variable which is only used once. also, name the variable after the input name, which considering the suggestion to change it, should be staleDays

```suggestion
    const staleDays = Number(core.getInput('stale-days'));
```

File: src/index.js [hunk 28]
```diff
@@ -1,36 +1,57 @@
...
+    const latestRelease = await getLatestRelease(token);
+
+    if (!latestRelease) {
+      logInfo('Could not find latest release');
+      return;
```
Comment (simoneb) [inline src/index.js:15]: minor:

```suggestion
      return logInfo('Could not find latest release');
```

**Review by simoneb — COMMENTED** [review]
See inline comments. Still some things to change but getting there.

File: package.json [hunk 29]
```diff
@@ -5,26 +5,35 @@
...
     "build": "ncc build src --license licenses.txt",
-    "lint": "eslint ."
+    "test": "jest --coverage"
   },
   "repository": {
     "type": "git",
-    "url": "git+https://github.com/estherixz/github-action-notify-release.git"
+    "url": "git+https://github.com/nearform/github-action-notify-release.git"
   },
   "author": "Esther Saimpou",
```
Comment (simoneb) [inline package.json:14]: please add yourself as a contributor https://docs.npmjs.com/cli/v7/configuring-npm/package-json#people-fields-author-contributors.

File: .github/workflows/ci.yml [hunk 30]
```diff
@@ -0,0 +1,21 @@
...
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Check out Git repository
+        uses: actions/checkout@v2
+
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v1
```
Comment (simoneb) [inline .github/workflows/ci.yml:14]: the latest version of the action is v2. can you please also add a dependabot.yml configuration so we keep the dependencies up to date? and please add our own automerge action. inspire from how the get-jwks build is done for instance

File: package.json [hunk 31]
```diff
@@ -5,26 +5,37 @@
   "main": "dist/index.js",
   "scripts": {
     "build": "ncc build src --license licenses.txt",
-    "lint": "eslint ."
+    "test": "jest --config=jest.config.js",
```
Comment (simoneb) [inline package.json:8]: no need to provide the configuration name if it's the default name

File: src/utils.js [hunk 32]
```diff
@@ -0,0 +1,49 @@
...
+  const octokit = github.getOctokit(token);
+
+  return octokit.issues.create({
+    ...github.context.repo,
+    title: issueTitle,
+    body: issueBody,
+  });
+}
+
+async function getLastOpenPendingIssue(token, latestReleaseDate) {
```
Comment (simoneb) [inline src/utils.js:14]: This logic load the existing issue doesn't look very robust. We should probably use the title of the issue instead of the author. The author is very generic here and I guess another action may have created an issue, meaning that we may be modifying an existing, unrelated issue. Let's use the issue title.

**Review by simoneb — CHANGES_REQUESTED** [review]
See inline comments. Please get used to adding a blank line to the end of all files.

File: jest.config.js [hunk 33]
```diff
@@ -0,0 +1,15 @@
+'use strict'
+
+module.exports = {
+  testEnvironment: 'node',
+  collectCoverage: true,
+  coverageDirectory: '/test/',
```
Comment (simoneb) [inline jest.config.js:6]: do not override this, let's use the default name for the coverage directory

File: jest.config.js [hunk 34]
```diff
@@ -0,0 +1,15 @@
...
+  "coverageThreshold": {
```
Comment (simoneb) [inline jest.config.js:7]: remove double quotes, keep consistency

File: src/utils.js (hunk 32 above)
Comment (sameer-coder) [reply src/utils.js:14]: The API doesn't allow us to use the issue title. There is no parameter called title which we can pass to the API. 
Maybe what we can do is set a `label` when creating the issue and then search for the issue using that label to update it. What do you think?

Comment (simoneb) [reply src/utils.js:14]: Ah interesting. Yes let's use something we can customize, a label then. does `notify-release` sound like a sensible label name?

Comment (sameer-coder) [reply src/utils.js:14]: Yep. `notify-release` should be good. I will make the changes

File: .github/workflows/ci.yml [hunk 35]
```diff
@@ -0,0 +1,21 @@
...
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Check out Git repository
+        uses: actions/checkout@v2
+
+      - name: Use Node.js ${{ matrix.node-version }}
```
Comment (simoneb) [inline .github/workflows/ci.yml:13]: you're not using matrix, remove the variable referencing it. note that you can omit the name field altogether, I would, it doesn't provide any additional useful information (here and in the other actions too)

File: .github/workflows/ci.yml [hunk 36]
```diff
@@ -0,0 +1,21 @@
...
+        uses: actions/setup-node@v2
+        with:
+          node-version: 12
```
Comment (simoneb) [inline .github/workflows/ci.yml:16]: let's use 14 which is LTS now

File: .github/workflows/ci.yml [hunk 37]
```diff
@@ -0,0 +1,21 @@
+name: CI
+
+on: pull_request
```
Comment (simoneb) [inline .github/workflows/ci.yml:3]:
```suggestion
on: [push, pull_request]
```

File: .husky/.gitignore [hunk 38]
```diff
@@ -0,0 +1 @@
+_
```
Comment (simoneb) [inline .husky/.gitignore:1]: what's this?

File: .github/workflows/ci.yml (hunk 36 above)
Comment (sameer-coder) [reply .github/workflows/ci.yml:16]: 14 is not available for use in github actions at the moment

File: .github/workflows/ci.yml (hunk 35 above)
Comment (sameer-coder) [reply .github/workflows/ci.yml:13]: OK

File: .husky/.gitignore (hunk 38 above)
Comment (sameer-coder) [reply .husky/.gitignore:1]: Its a folder added automatically by husky when you run `husky install`

File: .github/workflows/ci.yml (hunk 36 above)
Comment (simoneb) [reply .github/workflows/ci.yml:16]: who told you that? :D

File: .husky/.gitignore (hunk 38 above)
Comment (simoneb) [reply .husky/.gitignore:1]: fine then!

File: .github/workflows/ci.yml (hunk 36 above)
Comment (sameer-coder) [reply .github/workflows/ci.yml:16]: I got an error earlier when I tried to change it to node 14. Anyways I will check the docs

Comment (sameer-coder) [reply .github/workflows/ci.yml:16]: Strange! Docs mention  node-version: [10.x, 12.x, 14.x, 15.x] 
I will change it

Comment (sameer-coder) [reply .github/workflows/ci.yml]: I got this error after changing to node14 and running the action
`Error: System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter ''using: node14' is not supported, use 'docker' or 'node12' instead.')`

Comment (simoneb) [reply .github/workflows/ci.yml]: https://github.com/fastify/dependabot-merge-action-app/blob/master/.github/workflows/ci.yml#L10

File: .github/workflows/ci.yml [hunk 39]
```diff
@@ -0,0 +1,16 @@
...
+    runs-on: ubuntu-latest
+
+    steps:
+        uses: actions/checkout@v2
+        uses: actions/setup-node@v2
+        with:
+          node-version: 14
+        run: npm ci
+        run: npm run lint
+        run: npm test
```
Comment (simoneb) [inline .github/workflows/ci.yml:16]: please add dependabot automerge https://github.com/fastify/dependabot-merge-action-app/blob/master/.github/workflows/ci.yml#L15

File: action.yml [hunk 40]
```diff
@@ -1,13 +1,13 @@
...
-  days-to-stale-release:
-    description: 'The number of days after which notification should trigger after last unreleased commit'
-    required: true
+  stale-days:
+    description: 'Number of days of inactivity before a release becomes stale'
+    required: false
     default: 7
 runs:
-  using: 'node12'
+  using: 'node14'
```
Comment (simoneb) [inline action.yml:12]: this is the problem with node14. Here it is not supported, in CI it is supported.

File: jest.config.js [hunk 41]
```diff
@@ -0,0 +1,14 @@
+'use strict'
+
+module.exports = {
+  testEnvironment: 'node',
+  collectCoverage: true,
+  'coverageThreshold': {
```
Comment (simoneb) [inline jest.config.js:6]: remove quotes

File: jest.config.js [hunk 42]
```diff
@@ -0,0 +1,14 @@
...
+    'global': {
```
Comment (simoneb) [inline jest.config.js:7]: remove quotes

File: package.json [hunk 43]
```diff
@@ -5,26 +5,37 @@
...
   "devDependencies": {
     "@vercel/ncc": "^0.28.3",
-    "eslint": "^7.24.0"
+    "eslint": "^7.25.0",
+    "husky": "^6.0.0",
+    "jest": "^26.6.3"
+  },
+  "husky": {
+    "hooks": {
+      "pre-commit": "npm test && npm run lint && npm run build && git add dist"
```
Comment (simoneb) [inline package.json:38]: remove test and lint, we don't need them as they unnecessarily slow down the commits. we only want to build

Comment (sameer-coder) [reply package.json]: Done

**Review by simoneb — APPROVED** [review]
LGTM. Let's:

- merge
- do a release v1 from the master branch
- publish it to the marketplace
- set it up in get-jwks for a starter
